JSON Schema Generator

Michael7371 opened the pull request:
Adding JSON schema generation for J2735 messages based off of POJO definitions

iyourshaw committed:
json schema generator project

iyourshaw committed:
Custom schema generation for ASN.1 integer types with min and max

iyourshaw committed:
Custom providers for bitstrings, enums, boolean, OIDs

iyourshaw committed:
readme

iyourshaw committed:
clean

iyourshaw committed:
Fix raw ia5string type

iyourshaw committed:
Merge branch 'develop' into feature/jsonschema-generator

Michael7371 committed:
working unit tests and support for `Asn1ParameterizedTypes` extraction

Michael7371 committed:
updates to support `@Asn1Property` annotations

Michael7371 committed:
update to handle both static and variable length bit strings

Michael7371 committed:
Add support for ASN.1 choice types and octet strings in JSON schema generation

- Introduced handling for ASN.1 CHOICE types, allowing for union of possible types in schema generation.
- Added support for ASN.1 OCTET STRING types, including min and max length constraints.
- Enhanced the JsonSchemaGenerator to process nested definitions and merge them into the root schema.
- Updated unit tests to verify the presence of properties or oneOf in generated schemas.

Michael7371 committed:
Enhance JSON schema generation with new schemas and examples

- Added new JSON schema files for various ASN.1 message types including Basic Safety Message, Road Safety Message, and Sensor Data Sharing Message.
- Introduced multiple examples for each message type to demonstrate usage and validation.
- Updated .gitignore to exclude VSCode settings.
- Added a new shell script for batch schema generation.
- Enhanced README documentation for better clarity on usage and prerequisites.

Michael7371 committed:
removal of old sdsm schema

Michael7371 committed:
updating `provideParameterizedTypeDefinition` handler to generate value schemas recursively

Michael7371 committed:
Fixing sequence type schema generation

Michael7371 committed:
Refactor JSON schema files to remove unnecessary $schema references

- Removed redundant "$schema" properties from multiple schema files to streamline the JSON schema structure.
- Updated relevant properties to use "const" instead of "enum" for fixed values in schemas.
- Ensured consistency across various ASN.1 message schemas, enhancing clarity and maintainability.

Michael7371 committed:
removing extra logging and adding message frame schemas

Michael7371 committed:
Removing required fields that are extension types

iyourshaw reviewed (commented):
The updates look great.

There is one additional issue I see: with the message frame schemas.  It looks like they don't have the messageId, probabably because they were generated from the classes like "MapMessageFrame", which are a bit odd because they inherit from the generic MessageFrame<> class.  Those are actually parameterized so would require special handling.
 
But an easier way to get the Message Frame schemas would be to extract them from the "MessageFrame.schema.json" which is already there.  Since it is a big "oneOf" with all the message frames with the correct format, you could extract the ones of interest from it.  
 
Would be good to add to tests for the message frame schemas too (there are existing test message frame jsons in the pojos tests)

Michael7371 committed:
removal of message frame generation per type due to missing message ID

Michael7371 committed:
Add BasicSafetyMessageMessageFrame schema and update README

- Introduced a new JSON schema for BasicSafetyMessageMessageFrame to define its structure and properties.
- Updated README to include information on generating specific MessageFrame type schemas from the new schema.
- Clarified the placement of generated schemas in the schemas directory.

Michael7371 committed:
formatting sample BSM files

Michael7371 committed:
update axle value to be allowed by json schema definition

drewjj committed:
Create JSON schemas for all supported and planned to support jpo-ode message types message frames

drewjj committed:
Add tests for message frames, relocate schemas to resources for testing

drewjj commented:
> The updates look great.
> 
> There is one additional issue I see: with the message frame schemas. It looks like they don't have the messageId, probabably because they were generated from the classes like "MapMessageFrame", which are a bit odd because they inherit from the generic MessageFrame<> class. Those are actually parameterized so would require special handling.
> 
> But an easier way to get the Message Frame schemas would be to extract them from the "MessageFrame.schema.json" which is already there. Since it is a big "oneOf" with all the message frames with the correct format, you could extract the ones of interest from it.
> 
> Would be good to add to tests for the message frame schemas too (there are existing test message frame jsons in the pojos tests)

Ready for review again. All jpo-ode supported message frame schemas have individual tests for them.

Michael7371 committed:
removed generated message frame schema

Michael7371 committed:
Merge branch 'feature/jsonschema-generator' of https://github.com/CDOT-CV/jpo-asn-pojos into feature/jsonschema-generator

iyourshaw commented:
issue: just seeing a couple failed unit tests:

```
[Test worker] INFO us.dot.its.jpo.asn.jsonschema.generator.JsonSchemaGeneratorTest - Testing schema generation for PDU: basicSafetyMessage

Sample JSON should be valid against the generated schema: /us/dot/its/jpo/asn/jsonschema/generator/BasicSafetyMessage/message_frame/bsm_mf_allext_01.json
Expected: an empty collection
     but: <[$: required property 'coreData' not found]>
java.lang.AssertionError: Sample JSON should be valid against the generated schema: /us/dot/its/jpo/asn/jsonschema/generator/BasicSafetyMessage/message_frame/bsm_mf_allext_01.json
Expected: an empty collection
     but: <[$: required property 'coreData' not found]>
```

and

```
[Test worker] INFO us.dot.its.jpo.asn.jsonschema.generator.JsonSchemaGeneratorTest - Testing BasicSafetyMessageMessageFrame schema validation

Schema should have BasicSafetyMessage property
Expected: is <true>
     but: was <false>
java.lang.AssertionError: Schema should have BasicSafetyMessage property
Expected: is <true>
     but: was <false>
```

iyourshaw commented on jpo-asn-jsonschema-generator/batch_gen_schemas.sh:
```diff
@@ -0,0 +1,38 @@
+#!/bin/bash
+
+# Create output directory if it doesn't exist
+mkdir -p schemas
+
+# Check if the JAR file exists
+JAR_FILE="build/libs/schemagen-cli.jar"
+if [ ! -f "$JAR_FILE" ]; then
+    echo "Error: JAR file not found at $JAR_FILE"
+    echo "Please build the project first using: ./gradlew build"
+    exit 1
+fi
+
+# List of PDUs and their corresponding modules to generate schemas for
+# Format: "PDU_NAME:MODULE_NAME"
+PDUS=(
+    "MessageFrame:MessageFrame"
+    "BasicSafetyMessage:BasicSafetyMessage"
+    "PersonalSafetyMessage:PersonalSafetyMessage"
+    "SignalRequestMessage:SignalRequestMessage"
+    "SignalStatusMessage:SignalStatusMessage"
+    "SPAT:SPAT"
+    "MapData:MapData"
+    "SensorDataSharingMessage:SensorDataSharingMessage"
+    "RTCMcorrections:RTCMcorrections"
+    "RoadSafetyMessage:RoadSafetyMessage"
+)
+
+# Generate schema for each PDU
+for pdu_entry in "${PDUS[@]}"; do
+    # Split the entry into PDU and module names
+    IFS=':' read -r pdu module <<< "$pdu_entry"
+    echo "Generating schema for $pdu (module: $module)..."
+    mkdir -p schemas/${module}
+    java -jar build/libs/schemagen-cli.jar -m "$module" -p "$pdu" -o "schemas/${module}/${pdu}.schema.json"
+done
+
+echo "Schema generation complete!"
```
Suggestion: add the outputs of this script to .gitignore so we don't accidentally commit redundant copies of the schema

drewjj replied:
Added the output directory and all content to the .gitignore

iyourshaw commented:
one more suggestion: add jacoco and add the generator project to the CI workflow, so test results will show up in the PR

drewjj committed:
Add Jacoco

drewjj committed:
Add to GitHub CI

drewjj committed:
Add schemas output to the gitignore

drewjj committed:
Merge branch 'feature/jsonschema-generator' of https://github.com/CDOT-CV/jpo-asn-pojos into feature/jsonschema-generator

drewjj commented:
> one more suggestion: add jacoco and add the generator project to the CI workflow, so test results will show up in the PR

Added Jacoco and the output to the CI

Michael7371 committed:
fixing unit tests

Michael7371 committed:
Merge branch 'feature/jsonschema-generator' of https://github.com/CDOT-CV/jpo-asn-pojos into feature/jsonschema-generator

Michael7371 committed:
updating README documentation

iyourshaw commented on .github/workflows/ci.yml:
```diff
@@ -43,6 +43,18 @@ jobs:
         name: coverage-report-test-generator
         path: jpo-asn-test-generator/build/reports/jacoco/test/html/
 
+    - name: Run Tests with Coverage (jpo-asn-jsonschema-generator)
+      working-directory: ./jpo-asn-jsonschema-generator
+      run: |
+        chmod +x gradlew
+        ./gradlew check
+      
+    - name: Upload coverage reports (jpo-asn-jsonschema-generator)
+      uses: actions/upload-artifact@v4
+      with:
+        name: coverage-report
+        path: jpo-asn-jsonschema-generator/build/reports/jacoco/testCodeCoverageReport/html/
+
     - name: Add Coverage PR Comment (jpo-asn-j2735-2024 and jpo-asn-runtime)
```
suggestion (non-blocking): also add the madrapps/jacoco-report action to post the jacoco report as PR comments

drewjj replied:
Added

iyourshaw reviewed (approved):
Looks good, unit tests all pass and the `batch_gen_schemas.sh` script works.
Just one suggestion to add the Jacoco comments action.

drewjj committed:
Add madrapps/jacoco-report to the CI
